src/cryptonote_core/service_node_list.cpp

```diff
@@ -321,9 +329,201 @@ namespace service_nodes
       LOG_PRINT_L1("Deregistration for service node: " << key);
     }
 
-
     m_rollback_events.push_back(std::unique_ptr<rollback_event>(new rollback_change(block_height, key, iter->second)));
     m_service_nodes_infos.erase(iter);
+
+    return true;
+  }
+
+  static uint64_t get_new_swarm_id(uint64_t seed)
+  {
+    std::mt19937_64 mersenne_twister(seed);
+    return uniform_distribution_portable(mersenne_twister, UINT64_MAX);
+  }
+
+  static std::vector<crypto::public_key> get_snodes_in_swarm(std::unordered_map<crypto::public_key, service_node_info>& infos, SwarmID sid)
+  {
+    std::vector<crypto::public_key> result;
+    for (const auto& entry : infos) {
+      if (entry.second.swarm_id == sid) {
+        result.push_back(entry.first);
+      }
+    }
+    return result;
+  }
+
+  static std::vector<SwarmID> get_all_swarms(const std::map<SwarmID, size_t>& swarm_sizes) {
+    std::vector<SwarmID> all_swarms;
+    all_swarms.reserve(swarm_sizes.size());
+    for (const auto& entry : swarm_sizes) {
+      all_swarms.push_back(entry.first);
+    }
+    return all_swarms;
+  }
+
+  static crypto::public_key dequeue_snode(std::vector<crypto::public_key>& buffer, std::mt19937_64& mt) {
+    const auto idx = uniform_distribution_portable(mt, buffer.size());
+    const auto sn_pk = buffer.at(idx);
+    buffer.erase(buffer.begin() + idx);
+    return sn_pk;
+  }
+
+  void service_node_list::update_swarms(uint64_t height) {
+
+    crypto::hash hash = m_blockchain.get_block_id_by_height(height);
+    uint64_t seed = 0;
+    std::memcpy(&seed, hash.data, sizeof(seed));
+    std::mt19937_64 mersenne_twister(seed);
+
+    std::map<SwarmID, size_t> swarm_sizes;
```

Comment:
std::map<SwarmID, std::vector<public_keys>> looks more useful here than jumping between this and all_swarms?

Reply:
That would be more clear, but then keeping this map up to date would be more expensive (as opposed to incrementing/decrementing counts) as we update swarm ids. Though we are not concerned about performance here.